<div style="color:#483838; padding:0px 10px; border-radius:5px; font-size:18px; text-align:center"><h1 style='margin:10px 5px'>Misiing Values and scikit-learn</h1>
</div>

- __01)What is scikit-learn__
- __02)import dataset__
- __03)Extarcting missing values__
- __04)Replacing missing valuse using Scikit-Learn.SimpleImputer:__
- __a)using mean__
- __b)using median strategy__
- __c)using constant strategy__

## 01)What is scikit-learn

- Simple and efficient tools for predictive data analysis
- Accessible to everybody, and reusable in various contexts
- Built on NumPy, SciPy, and matplotlib
- Open source, commercially usable - BSD license
- instead of filling nan values by using of coding we use this library.

#### install  scikit-learn:
$$pip~install~ -U~ scikit-learn$$

##### import scikit-learn in order to fill nan features with mean, median, constant values and ....

In [74]:
from  sklearn.impute import SimpleImputer

ref: https://scikit-learn.org/stable

## 02)impor dataset

In [75]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [76]:
data_raw  = pd.read_csv("G:\\Data_analyst\\05)data prepration\\heart.csv")
data = data_raw.copy()    #make a copy in order to compare data

## 03)Extarcting missing values

In [77]:
data.isnull().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

- There is no null or NAN value in data,but in Cholesterol and RestingBP columns we have 0 values which are our missed values, since they are not allowed to be 0 in real world.

In [78]:
data[data.Cholesterol == 0]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
293   65   M           ASY        115            0          0     Normal   
294   32   M            TA         95            0          1     Normal   
295   61   M           ASY        105            0          1     Normal   
296   50   M           ASY        145            0          1     Normal   
297   57   M           ASY        110            0          1         ST   
..   ...  ..           ...        ...          ...        ...        ...   
514   43   M           ASY        122            0          0     Normal   
515   63   M           NAP        130            0          1         ST   
518   48   M           NAP        102            0          1         ST   
535   56   M           ASY        130            0          0        LVH   
536   62   M           NAP        133            0          1         ST   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
293     93              Y      0.0     Flat             1  
294    127              N      0.7       Up             1  
295    110              Y      1.5       Up             1  
296    139              Y      0.7     Flat             1  
297    131              Y      1.4       Up             1  
..     ...            ...      ...      ...           ...  
514    120              N      0.5       Up             1  
515    160              N      3.0     Flat             0  
518    110              Y      1.0     Down             1  
535    122              Y      1.0     Flat             1  
536    119              Y      1.2     Flat             1  

[172 rows x 12 columns]

In [79]:
data[data.RestingBP == 0]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
449   55   M           NAP          0            0          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
449    155              N      1.5     Flat             1

- There is 172 zeros in cholestrol column and one 0 in RestingBP column and we should change them to null in order to be able to use them in SimpleImputer algorithm, since it does not know 0 as a missing value.

- We should replace 0 valuse here with null value, in order to be able to decide to replace the with a correct value. 

In [80]:
data.replace({"Cholesterol":0},np.nan,inplace=True)
data.replace({"RestingBP":0},np.nan,inplace=True)

In [81]:
data.isnull().sum()

Age                 0
Sex                 0
ChestPainType       0
RestingBP           1
Cholesterol       172
FastingBS           0
RestingECG          0
MaxHR               0
ExerciseAngina      0
Oldpeak             0
ST_Slope            0
HeartDisease        0
dtype: int64

## 04)Replacing missing valuse using Scikit-Learn.SimpleImputer

In [82]:
from sklearn.impute import SimpleImputer
#Using simpleImputer we can fill missing values cells using mean, median, most_frequent(mode) and constant strategies.

__a)using mean__

In [83]:
simple_imp = SimpleImputer(missing_values= np.nan , strategy="mean")  #making an object from simpleimputer
simple_imp.fit(data[["Cholesterol"]])                         #fit the model learn from "Cholesterol" and save it in simple_imp
data["Cholesterol"] = simple_imp.transform(data[["Cholesterol"]]) #transform implement the educated data       

##### Here our replacing missing values has been done.
- let's check the mean value and missing values.

- We can see that the mean in cholestrol column has been changed from 177 to 244

In [84]:
data.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  917.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.540894   244.635389    0.233115  136.809368   
std      9.432617   17.999749    53.318029    0.423046   25.460334   
min     28.000000   80.000000    85.000000    0.000000   60.000000   
25%     47.000000  120.000000   214.000000    0.000000  120.000000   
50%     54.000000  130.000000   244.635389    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

- The mean value for Cholesterol has been calculated as 244.635389, that we calculated exactly this number using python code before.

In [85]:
data.isnull().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         1
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

__b)using median strategy__

In [86]:
#making a copy of our raw data
data2 = data_raw.copy()
data2.replace({"Cholesterol":0},np.nan,inplace=True)
simple_imp2 = SimpleImputer(missing_values= np.nan , strategy="median")

In [87]:
data2["Cholesterol"] = simple_imp2.fit_transform(data2[["Cholesterol"]])
#here we have fitted and transformed data simeltaneously

In [88]:
data2.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.396514   243.204793    0.233115  136.809368   
std      9.432617   18.514154    53.401297    0.423046   25.460334   
min     28.000000    0.000000    85.000000    0.000000   60.000000   
25%     47.000000  120.000000   214.000000    0.000000  120.000000   
50%     54.000000  130.000000   237.000000    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

__b)using median strategy__

In [95]:
#making a copy of our raw data
data2 = data_raw.copy()
data2.replace({"Cholesterol":0},np.nan,inplace=True)
simple_imp2 = SimpleImputer(missing_values= np.nan , strategy="median")

In [96]:
data2["Cholesterol"] = simple_imp2.fit_transform(data2[["Cholesterol"]])
#here we have fitted and transformed data simeltaneously

In [97]:
data2.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.396514   243.204793    0.233115  136.809368   
std      9.432617   18.514154    53.401297    0.423046   25.460334   
min     28.000000    0.000000    85.000000    0.000000   60.000000   
25%     47.000000  120.000000   214.000000    0.000000  120.000000   
50%     54.000000  130.000000   237.000000    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

__c)using constant strategy__

In [98]:
#making a copy of our raw data
data3 = data_raw.copy()
data3.replace({"Cholesterol":0},np.nan,inplace=True)
simple_imp3 = SimpleImputer(missing_values= np.nan , strategy="constant", fill_value=150)

In [99]:
data3["Cholesterol"] = simple_imp3.fit_transform(data3[["Cholesterol"]])
#here we have fitted and transformed data simeltaneously

In [100]:
data3.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.396514   226.904139    0.233115  136.809368   
std      9.432617   18.514154    64.868372    0.423046   25.460334   
min     28.000000    0.000000    85.000000    0.000000   60.000000   
25%     47.000000  120.000000   173.250000    0.000000  120.000000   
50%     54.000000  130.000000   223.000000    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

- Note that we cannot implement KNN to fill nan data here, since it use all data and we have some catogorical values here.So, we should chane categorical value to numerical values and then implement KNN.we learn to do that in next note book.